### PR TITLE
Implement node_match()

### PR DESCRIPTION
**Pull request by lionel-**

### CODE REVIEW

**Comment by lionel-** [comment]
Depends on tidyverse/rlang#270

Syntax and semantics described here: https://github.com/lionel-/redpen/pull/1/files#diff-15de555e9934617c671137c0181abcaa

cc @hadley

File: R/node-match.R [hunk 1]
```diff
@@ -0,0 +1,454 @@
+#' Match patterns to a call
+#'
+#' @description
```
Comment (hadley) [inline R/node-match.R:3]: This is probably too long for a description and should go in details or separate sections.

File: R/node-match.R [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+#'
+#' `node_match()` is a handy way of checking whether a quoted call
+#' conforms to a pattern. The following rules apply to patterns:
```
Comment (hadley) [inline R/node-match.R]: "The following rules apply to patterns:" -> "It uses the following rules"

File: R/node-match.R [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+#'
+#' * Named arguments. They will match even if not in order. The
```
Comment (hadley) [inline R/node-match.R]: Maybe `__Named__ arguments will match regardless of order`

File: R/node-match.R [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+#' * NULL-named arguments. A NULL name signals an unnamed argument
+#'   that should match even if not in order. The patterns `call(NULL =
+#'   foo, NULL = bar)` and `call(NULL = bar, NULL = foo)` will both
+#'   match the call `call(foo, bar)`.
+#'
+#'
+#' In addition, patterns can contain wildcards that will match
+#' anything:
+#'
+#' * The `.` wildcard will match any argument. It can appear on the
```
Comment (hadley) [inline R/node-match.R]: Would match any call with a single named argument?

File: R/node-match.R [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+#'   LHS or the RHS of an argument. For instance `call(. = .)` will
+#'   match any argument no matter its name.
+#'
+#' * The `...` wildcard matches all remaining unmatched arguments.
```
Comment (hadley) [inline R/node-match.R]: Maybe emphasise that the default matching is exhaustive? `i.e. f(x = 1, y = 1)` won't match `f(x = .)`

File: R/node-match.R [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+#'   `call(...)` will match any calls to `call()`, including
+#'   `call(foo)` or `call(foo, bar())`.
+#'
+#'   If you need to match an argument regardless of whether it has a
+#'   name, use the ellipsis as LHS: `call(... = .)`.
+#'
+#' * Binding wildcards with `.(name)`. Binding wildcards match
```
Comment (hadley) [inline R/node-match.R:55]: What happens if names aren't unique?

File: R/node-match.R [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+#' * Eval-binding wildcards with `..(name)`. These wildcards work just
+#'   like binding wildcards but evaluate the matched code before
+#'   binding it to a symbol.
+#'
+#' Patterns are supplied as formulas with the pattern on the left-hand
+#' side and an expression of your choosing on the right-hand side. The
+#' expression is evaluated only if the pattern match. It can evaluate
+#' to a sentinel value (so you know which expression matched) or to
+#' some checking code. If no pattern matches, the return value of is
+#' `NULL`.
```
Comment (hadley) [inline R/node-match.R]: Needs small example here. Maybe this would be better as vignette?

**Review by hadley — COMMENTED** [review]
Have just reviewed the documentation - I like that you included a little guidance in the review so I can start with the docs to get an overview.

Looks great so far.

File: R/node-match.R (hunk 6 above)
Comment (lionel-) [reply R/node-match.R:55]: The bindings will just be overridden. We could check whether a binding already exists and issue a warning or an error?

Comment (hadley) [reply R/node-match.R:55]: I don't think we need to change the behaviour, just document it.

File: R/node-match.R (hunk 2 above)
Comment (garrettgman) [reply R/node-match.R]: I think the new reader will become confused at this point if you do not add a more complete description of node_match before you move into pattern rules, e.g. 

"`node_match()` is a handy way of checking whether a quoted call conforms to a pattern. `node_match()` compares the call to a set of pattern ~ expression pairs. It evaluates the expression portion of the first pair whose pattern portion matches the call. 

`node_match()` uses the following rules to match call arguments to patterns..."

File: R/node-match.R [hunk 8]
```diff
@@ -0,0 +1,454 @@
+#' Match patterns to a call
```
Comment (garrettgman) [inline R/node-match.R:1]: I think that it is important to choose names that map as closely as possible to the teacher's mental model of grading. At first glance, `node_match()` doesn't seem to do that; it seems like something that might make more sense to a programmer than a teacher. On the other hand `node_match()` is a great name if we are educating users to think of each piece of student code as a network with nodes, for example to make it easier to navigate and isolate sections of the code. If this is the case, ignore my comment. I think it is fair to ask users to adopt new vocabulary so long as 1) there is a good reason for it, and 2) we teach the vocabulary and the bigger picture that it ties in to in a vignette.

If this was an arbitrary choice, I suggest something like code_match, call_match, function_match, pattern_match, etc.

Comment (lionel-) [reply R/node-match.R:1]: Agreed. `node_match()` as a name makes more sense within the landscape of low-level tools provided in rlang. It is still unclear the place `node_match()` will have in the final redpen UI, maybe we'll have a higher-level wrapper. If the teachers are going to use `node_match()`, we'll have to think about either providing a less technical-sounding synonym for it, or teach them about the tree structure of language calls as you suggest.

Comment (hadley) [reply R/node-match.R:1]: FWIW I think teachers are going to have to internalise the tree model of code, at least to the level described in https://adv-r.hadley.nz/expressions.html

**Comment by garrettgman** [comment]
Agreed. I think we are coming up with a useful mental model of code as much
as a package. We should think about how to deliver both.

**Pull request merged**